hazelcast/src/main/java/com/hazelcast/client/impl/spi/ProxyManager.java

```diff
@@ -283,39 +283,47 @@ private ClientProxy getOrCreateProxyInternal(@Nonnull String service,
         checkNotNull(service, "Service name is required!");
         checkNotNull(id, "Object name is required!");
 
+        final ObjectNamespace ns = new DistributedObjectNamespace(service, id);
+        ClientProxyFuture proxyFuture;
+
         distributedObjectsReadLock.lock();
         try {
-            final ObjectNamespace ns = new DistributedObjectNamespace(service, id);
-            ClientProxyFuture proxyFuture = proxies.get(ns);
+            proxyFuture = proxies.get(ns);
             if (proxyFuture != null) {
                 return proxyFuture.get();
             }
-            ClientProxyFactory factory = proxyFactories.get(service);
-            if (factory == null) {
-                throw new ClientServiceNotFoundException("No factory registered for service: " + service);
-            }
+
+        } finally {
+            distributedObjectsReadLock.unlock();
+        }
+
+        ClientProxyFactory factory = proxyFactories.get(service);
+        if (factory == null) {
+            throw new ClientServiceNotFoundException("No factory registered for service: " + service);
+        }
+
+        // We can do a write operation in the following block, thus we use the write lock.
+        distributedObjectsWriteLock.lock();
+        try {
             proxyFuture = new ClientProxyFuture();
```

Comment:
I think, you need a [double checked lock idiom](https://en.wikipedia.org/wiki/Double-checked_locking) here. check again under the write lock.

Reply:
locking mechanism changed.